### PR TITLE
[flutter_tool] Refactor Logger and Terminal to (mostly) no longer depend directly on context

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

The Logger (and consequently Terminal) is used in almost every part of the flutter code base. In order to support context-free testing, this class must be refactored before others.

The changes here mostly involve taking what were context values and providing them through the constructor. There is still a bit of weirdness with the Status' classes Stopwatch injection that I don't yet have a good answer for.

Finally, I cleaned up several tests that were unnecessarily creating extra Loggers

Followup enhancements to this would be to create a "BufferLogger.test()" factory that produced a preconfigured buffer logger without requiring every argument.

https://github.com/flutter/flutter/issues/47161

### CODE REVIEW

File: packages/flutter_tools/test/general.shard/base/logger_test.dart [hunk 1]
```diff
@@ -30,8 +30,14 @@ void main() {
       fakeStopWatch = FakeStopwatch();
     });
     testUsingContext('error', () async {
-      final BufferLogger mockLogger = BufferLogger();
-      final VerboseLogger verboseLogger = VerboseLogger(mockLogger);
+      final BufferLogger mockLogger = BufferLogger(
+        terminal: AnsiTerminal(
+          stdio: stdio, // Danger, using real stdio
```
Comment (zanderso) [inline packages/flutter_tools/test/general.shard/base/logger_test.dart]: Maybe also explain the reason for using the real stdio.

**Review by zanderso — COMMENTED** [review]
More for the globals change, it may be good to have a test like forbidden_imports_test for the sources/libraries that shouldn't be importing globals.dart. It could start out as a black list then flip to a whitelist.

File: dev/devicelab/bin/tasks/technical_debt__cost.dart [hunk 2]
```diff
@@ -137,6 +137,7 @@ Future<void> main() async {
         _kCostBenchmarkKey,
         _kNumberOfDependenciesKey,
         _kNumberOfConsumerDependenciesKey,
+        _kNumberOfFlutterToolGlobals,
```
Comment (jonahwilliams) [inline dev/devicelab/bin/tasks/technical_debt__cost.dart]: I forgot to add this key, so the benchmark isn't showing up 😢

File: packages/flutter_tools/lib/src/runner/flutter_command_runner.dart [hunk 3]
```diff
@@ -147,13 +147,16 @@ class FlutterCommandRunner extends CommandRunner<void> {
...
+      columnWidth: outputPreferences.wrapColumn,
+      shouldWrap: outputPreferences.wrapText,
+    );
   }
 
   @override
   String get usage {
     final String usageWithoutDescription = super.usage.substring(description.length + 2);
-    return  '${wrapText(description)}\n\n$usageWithoutDescription';
+    return '${wrapText(description, shouldWrap: outputPreferences.wrapText, columnWidth: outputPreferences.wrapColumn)}\n\n$usageWithoutDescription';
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/runner/flutter_command_runner.dart]: Fixed

**Comment by jonahwilliams** [comment]
I've rebased this on top of the prior globals change

File: dev/devicelab/bin/tasks/technical_debt__cost.dart (hunk 2 above)
Comment (zanderso) [reply dev/devicelab/bin/tasks/technical_debt__cost.dart]: Do you want to land adding the key in a separate PR to get initial numbers first before landing this?

File: packages/flutter_tools/lib/src/runner/flutter_command_runner.dart [hunk 4]
```diff
@@ -146,13 +145,20 @@ class FlutterCommandRunner extends CommandRunner<void> {
 
   @override
   String get usageFooter {
-    return wrapText('Run "flutter help -v" for verbose help output, including less commonly used options.');
+    return wrapText('Run "flutter help -v" for verbose help output, including less commonly used options.',
+      columnWidth: outputPreferences.wrapColumn,
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/runner/flutter_command_runner.dart]: Should `wrapText` take a single named arg for the output preferences instead?

File: packages/flutter_tools/test/general.shard/base/logger_test.dart [hunk 5]
```diff
@@ -31,8 +30,14 @@ void main() {
       fakeStopWatch = FakeStopwatch();
     });
     testUsingContext('error', () async {
-      final BufferLogger mockLogger = BufferLogger();
-      final VerboseLogger verboseLogger = VerboseLogger(mockLogger);
+      final BufferLogger mockLogger = BufferLogger(
```
Comment (zanderso) [inline packages/flutter_tools/test/general.shard/base/logger_test.dart]: Does this test still need a context?

File: dev/devicelab/bin/tasks/technical_debt__cost.dart (hunk 2 above)
Comment (jonahwilliams) [reply dev/devicelab/bin/tasks/technical_debt__cost.dart]: Good idea, opened here: https://github.com/flutter/flutter/pull/48266

File: packages/flutter_tools/test/general.shard/base/logger_test.dart (hunk 5 above)
Comment (jonahwilliams) [reply packages/flutter_tools/test/general.shard/base/logger_test.dart]: Fixed

File: packages/flutter_tools/lib/src/runner/flutter_command_runner.dart (hunk 4 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/runner/flutter_command_runner.dart]: There are a few properties that don't come from output preferences. I think the API needs some refactoring, but its used quite a bit so I'd rather do that in a different PR.

File: dev/devicelab/bin/tasks/technical_debt__cost.dart (hunk 2 above)
Comment (jonahwilliams) [reply dev/devicelab/bin/tasks/technical_debt__cost.dart]: Fixed

File: packages/flutter_tools/test/general.shard/base/logger_test.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/test/general.shard/base/logger_test.dart]: Fixed by switching to mock stdio

File: packages/flutter_tools/test/general.shard/base/logger_test.dart [hunk 6]
```diff
@@ -669,50 +736,73 @@ void main() {
...
-    }, overrides: <Type, Generator>{Stdio: () => mockStdio, Platform: _kNoAnsiPlatform});
+    }, overrides: <Type, Generator>{
+      Platform: () => _kNoAnsiPlatform,
+    });
 
-    testUsingContext('sequential startProgress calls with StdoutLogger', () async {
-      final Logger logger = context.get<Logger>();
-      logger.startProgress('AAA', timeout: timeoutConfiguration.fastOperation)..stop();
-      logger.startProgress('BBB', timeout: timeoutConfiguration.fastOperation)..stop();
+    // Status still uses Stopwatch from context.
```
Comment (zanderso) [inline packages/flutter_tools/test/general.shard/base/logger_test.dart]: TODO(jonahwilliams): Change to testWithoutContext when Status no longer uses the Stopwatch from the context.

**Review by zanderso — APPROVED** [review]

**Pull request merged**